File: app/src/main/java/dev/ragnarok/fenrir/api/adapters/CustomCommentsResponseAdapter.java
package dev.ragnarok.fenrir.api.adapters;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;

import dev.ragnarok.fenrir.api.model.response.CustomCommentsResponse;

public class CustomCommentsResponseAdapter extends AbsAdapter implements JsonDeserializer<CustomCommentsResponse> {
    private static final String TAG = CustomCommentsResponseAdapter.class.getSimpleName();

    @Override
    public CustomCommentsResponse deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        if (!checkObject(json)) {
            throw new JsonParseException(TAG + " error parse object");
        }
        CustomCommentsResponse response = new CustomCommentsResponse();
        JsonObject root = json.getAsJsonObject();

        JsonElement main = root.get("main");
        if (checkObject(main)) {
            response.main = context.deserialize(main, CustomCommentsResponse.Main.class);
        } // "main": false (if has execute errors)

        if (root.has("first_id")) {
            JsonElement firstIdJson = root.get("first_id");
            response.firstId = firstIdJson.isJsonNull() ? null : firstIdJson.getAsInt();
        }

        if (root.has("last_id")) {
            JsonElement lastIdJson = root.get("last_id");
            response.lastId = lastIdJson.isJsonNull() ? null : lastIdJson.getAsInt();
        }

        if (root.has("admin_level")) {
            JsonElement adminLevelJson = root.get("admin_level");
            response.admin_level = adminLevelJson.isJsonNull() ? null : adminLevelJson.getAsInt();
        }

        return response;
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/api/adapters/CustomCommentsResponseAdapter.kt
package dev.ragnarok.fenrir.api.adapters

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import dev.ragnarok.fenrir.api.model.response.CustomCommentsResponse
import java.lang.reflect.Type

class CustomCommentsResponseAdapter : AbsAdapter(), JsonDeserializer<CustomCommentsResponse> {
    @Throws(JsonParseException::class)
    override fun deserialize(
        json: JsonElement,
        typeOfT: Type,
        context: JsonDeserializationContext
    ): CustomCommentsResponse {
        if (!checkObject(json)) {
            throw JsonParseException("$TAG error parse object")
        }
        val response = CustomCommentsResponse()
        val root = json.asJsonObject
        val main = root["main"]
        if (checkObject(main)) {
            response.main = context.deserialize(main, CustomCommentsResponse.Main::class.java)
        } // "main": false (if has execute errors)
        if (root.has("first_id")) {
            val firstIdJson = root["first_id"]
            response.firstId = if (firstIdJson.isJsonNull) null else firstIdJson.asInt
        }
        if (root.has("last_id")) {
            val lastIdJson = root["last_id"]
            response.lastId = if (lastIdJson.isJsonNull) null else lastIdJson.asInt
        }
        if (root.has("admin_level")) {
            val adminLevelJson = root["admin_level"]
            response.admin_level = if (adminLevelJson.isJsonNull) null else adminLevelJson.asInt
        }
        return response
    }

    companion object {
        private val TAG = CustomCommentsResponseAdapter::class.java.simpleName
    }
}